numbat-wasm/www/service-worker.js

```diff
@@ -0,0 +1,181 @@
+// The prefix is necessary if multiple web apps will be hosted at the same
+// origin, e.g., as would be done on GitHub Pages. This prevents separate apps
+// from clashing when it comes to origin-scoped resources such as Caches, OPFS,
+// etc.
+const APP_PREFIX = "numbat_";
+// This should be updated any time cached assets change. It allows the entire
+// set of precached files to be refreshed atomically on SW update.
+// TODO: It would be nice if we did this automatically whenever a commit affects
+// cached files. Note that this could also be a sequence number or similar.
+// Random UUIDs just make it easier to avoid collisions without coordination.
+const VERSION_UUID = "819e71eb-c3df-46b8-885e-5650feaea241";
+const CACHE_NAME = `${APP_PREFIX}${VERSION_UUID}`;
+
+// NOTE: If these lists get very large, they should be moved into Sets for
+// membership tests.
+
+// Files atomically pre-cached at installation/update.
+// NOTE: We do _not_ cache the service worker itself.
+// TODO: Consider self-hosting Google fonts and caching. Alternatively, we could
+// cache the remote fonts directly.
+const PRECACHED_FILES = [
+  "./",
+  "index.html",
+  "index.js",
+  "jquery.min.js",
+  "jquery.mousewheel-min.js",
+  "jquery.terminal.min.js",
+  "keyboardevent-key-polyfill.js",
+  "main.css",
+  "opensearch.xml",
+  "pkg/numbat_wasm.js",
+  "pkg/numbat_wasm_bg.wasm",
+  "pkg/package.json", // Does this need to be cached?
+  "terminal.css",
+];
+// Files cached on-demand (at first load), but not refreshed. This should mostly
```

Comment:
Thanks for writing these comments, but I'm still not sure I understand the difference between PRECACHED and CACHED.

Reply:
`PRECACHED` files are downloaded deterministically before the first request to the service worker, since we know they will be used. `CACHED` files are cached on-demand at first request but then not updated thereafter (until the next app release, via service worker update). This helps save space because certain user agents will download different subset of optional assets, such as icons. This would be more pronounced if we have more icon sizes, but I've tested it in different browsers and they do in fact request different icon sizes and formats. There's no way to tell in advance which subset a given client might want.

I wasn't sure how to name this. It's not the same as "dynamically cached", since that refers to files whose contents are expected to change with every request (currently the exchange rate data).

Reply:
Agreed. Especially in the case of auto-generation, this is simpler and more robust if everything is precached.